Platform API changes for run image extension (Dockerfiles phase 3) (#347)
buildpacks/spec: platform.md

### Operations

#### Build

A single app image build* consists of the following phases:

1. Analysis
2. Detection
3. Cache Restoration
4. (Optional and Experimental) Base Image Extension
5. Build*

> 6. Export

Author: My IDE did this... I can put it back if desired

### Operations

#### Build

A single app image build* consists of the following phases:

1. Analysis
2. Detection
3. Cache Restoration
4. (Optional and Experimental) Base Image Extension
5. Build*
6. Export

A platform MUST execute these phases either by invoking the following phase-specific lifecycle binaries in order:
1. `/cnb/lifecycle/analyzer`
2. `/cnb/lifecycle/detector`
3. `/cnb/lifecycle/restorer`

> 4. (Optional and Experimental) `/cnb/lifecycle/extender`

Reviewer: ```suggestion
4. `/cnb/lifecycle/extender` (Optional and Experimental) 
```

Reviewer: do we define experimental anywhere?

- Fail if a rebase operation would remove OS packages.

#### Launch

> `/cnb/lifecycle/launcher` is responsible for launching user- and buildpack-provided processes in the correct execution environment.

Reviewer 2: is this `-` a mistake?

Author: It was intended as shorthand for "user-provided and buildpack-provided" ...I can update it to avoid any confusion

Author: Updated in b4f953901c717bfddcfe1e0398054920a109dd7f

> `/cnb/lifecycle/launcher` is responsible for launching user- and buildpack-provided processes in the correct execution environment.

Author: ```suggestion
`/cnb/lifecycle/launcher` is responsible for launching user-provided and buildpack-provided processes in the correct execution environment.
```

| `<analyzed>`    | `CNB_ANALYZED_PATH`  | `<layers>/analyzed.toml`                               | Path to output analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)                                                                                  |
| `<app>`         | `CNB_APP_DIR`        | `/workspace`                                           | Path to application directory                                                                                                                                |
| `<buildpacks>`  | `CNB_BUILDPACKS_DIR` | `/cnb/buildpacks`                                      | Path to buildpacks directory (see [Buildpacks Directory Layout](#buildpacks-directory-layout))                                                               |
| `<extensions>`^ | `CNB_EXTENSIONS_DIR` | `/cnb/extensions`                                      | (**[experimental](#experimental-features)**) Path to image extensions directory (see [Image Extensions Directory Layout](#image-extensions-directory-layout) |

> | `<generated>`^  | `CNB_GENERATED_DIR`  | `<layers>/generated`                                   | (**[experimental](#experimental-features)**) Path to output directory for generated Dockerfiles                                                              |

Author: Added footnotes to designate things that are needed for extensions only

- SHALL filter the build plan with dependencies provided by image extensions.
- SHALL copy any generated run.Dockerfiles to `<generated>/run/<image extension ID>/Dockerfile`.
- SHALL copy any generated build.Dockerfiles to `<generated>/build/<image extension ID>/Dockerfile`.
- SHALL copy any generated `<extend-config>` files to `<generated>/build/<image extension ID>/<extend-config>`.
- SHALL replace `run-image` in `<analyzed>` with the selected run image. To select the run image, the lifecycle SHALL inspect each `run.Dockerfile` output by image extensions, in the order defined in `<group>`:
  - **If** all `run.Dockerfile`s declare `FROM ${base_image}`, the selected run image SHALL be the original run image in `<analyzed>`, with `extend = true`
  - **Else** the selected run image SHALL be the last image referenced in the `FROM` statement of the last `run.Dockerfile` not to declare `FROM ${base_image}`
    - `run-image.image` SHALL be the name of the selected run image
    - `run-image.reference` and `run-image.target` SHALL be cleared (as they may no longer be accurate)

>     - All preceding `run.Dockerfile`s SHALL be renamed `run.Dockerfile.ignore`

Author: ```suggestion
    - All preceding `run.Dockerfile`s SHALL be copied to `<generated>` with suffix `.ignore`
```

    - **If** there are no `run.Dockerfile`s following the Dockerfile with the selected run image:
      - `run-image.extend` SHALL be `false`
    - **Else**
      - `run-image.extend` SHALL be `true`
- SHALL fail if the selected run image is not found in `<run>`

#### `restorer`

Usage:
```
/cnb/lifecycle/restorer \
  [-analyzed <analyzed>] \
  [-build-image <build-image>] \
  [-cache-dir <cache-dir>] \
  [-cache-image <cache-image>] \

>   [-generated <generated>] \

Author: ```suggestion
```

- SHALL copy any generated build.Dockerfiles to `<generated>/build/<image extension ID>/Dockerfile`.
- SHALL copy any generated `<extend-config>` files to `<generated>/build/<image extension ID>/<extend-config>`.
- SHALL replace `run-image` in `<analyzed>` with the selected run image. To select the run image, the lifecycle SHALL inspect each `run.Dockerfile` output by image extensions, in the order defined in `<group>`:
  - **If** all `run.Dockerfile`s declare `FROM ${base_image}`, the selected run image SHALL be the original run image in `<analyzed>`, with `extend = true`
  - **Else** the selected run image SHALL be the last image referenced in the `FROM` statement of the last `run.Dockerfile` not to declare `FROM ${base_image}`
    - `run-image.image` SHALL be the name of the selected run image
    - `run-image.reference` and `run-image.target` SHALL be cleared (as they may no longer be accurate)
    - All preceding `run.Dockerfile`s SHALL be renamed `run.Dockerfile.ignore`
    - **If** there are no `run.Dockerfile`s following the Dockerfile with the selected run image:
      - `run-image.extend` SHALL be `false`
    - **Else**
      - `run-image.extend` SHALL be `true`

> - SHALL fail if the selected run image is not found in `<run>`

Reviewer 3: This may be nicer if it doesn't fail outright, but is a warning, or at least could be reduced to a warning via a flag or env var. 

I'm thinking here that it may be required in some circumstances for customers to override the selected run image to one of their choosing (buildpacks supports this today for non-switched images, and arguably an extension may be able to make a more nuanced choice).. 

When combined with stack removal, which may make it a lot easier to use an image as a run image, it does make sense to have some way to override, although there are various options.. 

 - force override to a single given image
 - force override via a map of run.toml image -> desired image
 - force override to a constructed image reference using some rule applied via extensions (eg, baseimage:extension1-extension2-extension3)
 
 The 2nd & 3rd options although probably most powerful in conjunction with extension image switching also feel potentially too big to be containable within this item, but either way, would require the weakening of the 'SHALL' within the above statement.

Author: Changed to a warning

> - SHALL fail if the selected run image is not found in `<run>`

Author: ```suggestion
- SHALL warn if the selected run image is not found in `<run>`
```

> - SHALL fail if the selected run image is not found in `<run>`

Author: ```suggestion
- SHALL warn if the selected run image is not found in `<run>`
```

```

##### Inputs

| Input            | Environment Variable | Default Value            | Description                                                                                       |
|------------------|----------------------|--------------------------|---------------------------------------------------------------------------------------------------|
| `<analyzed>`     | `CNB_ANALYZED_PATH`  | `<layers>/analyzed.toml` | Path to output analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)                       |

> | `<build-image>`* | `CNB_BUILD_IMAGE`    |                          | Reference to the current build image in an OCI registry (if used `<kaniko-dir>` must be provided) |

Author: Added footnote

| `<cache-dir>`    | `CNB_CACHE_DIR`      |                          | Path to a cache directory                                                                         |
| `<cache-image>`  | `CNB_CACHE_IMAGE`    |                          | Reference to a cache image in an OCI registry                                                     |

> | `<generated>`^   | `CNB_GENERATED_DIR`  | `<layers>/generated`     | (**[experimental](#experimental-features)**) Path to directory containing generated Dockerfiles   |

Author: Added footnote

| `<gid>`          | `CNB_GROUP_ID`       |                          | Primary GID of the build image `User`                                                             |
| `<group>`        | `CNB_GROUP_PATH`     | `<layers>/group.toml`    | Path to group definition (see [`group.toml`](#grouptoml-toml))                                    |
| `<layers>`       | `CNB_LAYERS_DIR`     | `/layers`                | Path to layers directory                                                                          |
| `<log-level>`    | `CNB_LOG_LEVEL`      | `info`                   | Log Level                                                                                         |
| `<uid>`          | `CNB_USER_ID`        |                          | UID of the build image `User`                                                                     |
| `<skip-layers>`  | `CNB_SKIP_LAYERS`    | `false`                  | Do not perform [layer restoration](#layer-restoration)                                            |

> | `<kaniko-dir>`^  |                      |                          | Kaniko directory (must be `/kaniko`)                                                              |

Author: Added footnote

```

##### Inputs

| Input            | Environment Variable | Default Value            | Description                                                                                       |
|------------------|----------------------|--------------------------|---------------------------------------------------------------------------------------------------|
| `<analyzed>`     | `CNB_ANALYZED_PATH`  | `<layers>/analyzed.toml` | Path to output analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)                       |
| `<build-image>`* | `CNB_BUILD_IMAGE`    |                          | Reference to the current build image in an OCI registry (if used `<kaniko-dir>` must be provided) |
| `<cache-dir>`    | `CNB_CACHE_DIR`      |                          | Path to a cache directory                                                                         |
| `<cache-image>`  | `CNB_CACHE_IMAGE`    |                          | Reference to a cache image in an OCI registry                                                     |

> | `<generated>`^   | `CNB_GENERATED_DIR`  | `<layers>/generated`     | (**[experimental](#experimental-features)**) Path to directory containing generated Dockerfiles   |

Author: ```suggestion
```

- For each buildpack in `<group>`, if persistent metadata for that buildpack exists in the analysis metadata, lifecycle MUST write a toml representation of the persistent metadata to `<layers>/<buildpack-id>/store.toml`
- **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer restoration.
- **Else** the lifecycle MUST perform [layer restoration](#layer-restoration) for any app image layers or cached layers created by any buildpack present in the provided `<group>`.

> - When `<generated>/build` contains image extensions (optional and **[experimental](#experimental-features)**), the lifecycle:

Author: ```suggestion
- When `<generated>/build` contains Dockerfiles (optional and **[experimental](#experimental-features)**), the lifecycle:
```

  - MUST record the digest reference to the provided `<build-image>` in `<analyzed>`
  - MUST copy the OCI manifest and config file for `<build-image>` to `<kaniko-dir>/cache`

> - When `<generated>/run` contains image extensions, the lifecycle:

Author: ```suggestion
- When `<generated>/run` contains Dockerfiles, the lifecycle:
```

>   - MUST resolve `run-image.reference` to a digest in `<analyzed>`

Author: ```suggestion
  - MUST resolve `run-image.reference` to a digest in `<analyzed>` if not present
```

  - MUST populate `run-image.target` data in `<analyzed>` if not present
  - **If** `<analyzed>` has `run-image.extend = true`, the lifecycle:

>     - MUST copy in OCI layout format the `run-image` in `<analyzed>` to `<kaniko-dir>/cache`

Author: ```suggestion
    - MUST download from the registry and save in OCI layout format the `run-image` in `<analyzed>` to `<kaniko-dir>/cache`
```

- For each buildpack in `<group>`, if persistent metadata for that buildpack exists in the analysis metadata, lifecycle MUST write a toml representation of the persistent metadata to `<layers>/<buildpack-id>/store.toml`
- **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer restoration.
- **Else** the lifecycle MUST perform [layer restoration](#layer-restoration) for any app image layers or cached layers created by any buildpack present in the provided `<group>`.
- When `<generated>/build` contains Dockerfiles (optional and **[experimental](#experimental-features)**), the lifecycle:
  - MUST record the digest reference to the provided `<build-image>` in `<analyzed>`
  - MUST copy the OCI manifest and config file for `<build-image>` to `<kaniko-dir>/cache`

> - When `<generated>/run` contains Dockerfiles, the lifecycle:

Author: ```suggestion
- The lifecycle:
```

>   - MUST resolve `run-image.reference` to a digest in `<analyzed>` if not present

Author: ```suggestion
  - MUST resolve `run-image.reference` to a digest reference in `<analyzed>` if not present
```

- For each buildpack in `<group>`, if persistent metadata for that buildpack exists in the analysis metadata, lifecycle MUST write a toml representation of the persistent metadata to `<layers>/<buildpack-id>/store.toml`
- **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer restoration.
- **Else** the lifecycle MUST perform [layer restoration](#layer-restoration) for any app image layers or cached layers created by any buildpack present in the provided `<group>`.

> - When `<generated>/build` contains Dockerfiles (optional and **[experimental](#experimental-features)**), the lifecycle:

Author: ```suggestion
- When `<build-image>` is provided (optional and **[experimental](#experimental-features)**), the lifecycle:
```

| `1-10`, `13-19` | Generic lifecycle errors            |
| `100-109`       | Extension-specific lifecycle errors |

- For each extension in `<group>` in order, if a Dockerfile exists in `<generated>/<kind>/<buildpack-id>`, the lifecycle:

>   - SHALL apply the Dockerfile to the build environment according to the process outlined in the [Image Extension Specification](image-extension.md).

Author: ```suggestion
  - SHALL apply the Dockerfile to the environment according to the process outlined in the [Image Extension Specification](image-extension.md).
```

- The extended image MUST be an extension of:
  - The `build-image` in `<analyzed>` when `<kind>` is `build`, or
  - The `run-image` in `<analyzed>` when `<kind>` is `run`
- When extending the build image, after all `build.Dockefile`s are applied, the lifecycle:
  - SHALL proceed with the `build` phase using the provided `<gid>` and `<uid>`
- When extending the run image, after all `run.Dockefile`s are applied, the lifecycle:
  - SHALL copy the manifest and config for the extended run image, along with any new layers, to `<extended>`

>   - SHALL resolve `run-image.reference` in `<analyzed>` to the digest of the extended run image

Author: ```suggestion
```

| `1-10`, `13-19` | Generic lifecycle errors            |
| `100-109`       | Extension-specific lifecycle errors |

- For each extension in `<group>` in order, if a Dockerfile exists in `<generated>/<kind>/<buildpack-id>`, the lifecycle:

>   - SHALL apply the Dockerfile to the environment according to the process outlined in the [Image Extension Specification](image-extension.md).

Reviewer: `SHALL` is always odd to me. Does this mean it's required, like "MUST"?

Reviewer: or is it optional like "MAY"?

Author: I've had to look this up myself. It's required: 

[1](https://www.rfc-editor.org/rfc/rfc2119#section-1). MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the definition is an absolute requirement of the specification.

| `1-10`, `13-19` | Generic lifecycle errors            |
| `100-109`       | Extension-specific lifecycle errors |

- For each extension in `<group>` in order, if a Dockerfile exists in `<generated>/<kind>/<buildpack-id>`, the lifecycle:
  - SHALL apply the Dockerfile to the environment according to the process outlined in the [Image Extension Specification](image-extension.md).
- The extended image MUST be an extension of:
  - The `build-image` in `<analyzed>` when `<kind>` is `build`, or
  - The `run-image` in `<analyzed>` when `<kind>` is `run`
- When extending the build image, after all `build.Dockefile`s are applied, the lifecycle:
  - SHALL proceed with the `build` phase using the provided `<gid>` and `<uid>`
- When extending the run image, after all `run.Dockefile`s are applied, the lifecycle:
  - **If** any `run.Dockerfile` set the label `io.buildpacks.rebasable` to `false` or left the label unset:
    - SHALL set the label `io.buildpacks.rebasable` to `false` on the extended run image

>   - SHALL copy the manifest and config for the extended run image, along with any new layers, to `<extended>`

Author: ```suggestion
  - SHALL copy the manifest and config for the extended run image, along with any new layers, to `<extended>`/run
```

| `1-10`, `13-19` | Generic lifecycle errors            |
| `100-109`       | Extension-specific lifecycle errors |

- For each extension in `<group>` in order, if a Dockerfile exists in `<generated>/<kind>/<buildpack-id>`, the lifecycle:
  - SHALL apply the Dockerfile to the environment according to the process outlined in the [Image Extension Specification](image-extension.md).
- The extended image MUST be an extension of:
  - The `build-image` in `<analyzed>` when `<kind>` is `build`, or
  - The `run-image` in `<analyzed>` when `<kind>` is `run`
- When extending the build image, after all `build.Dockefile`s are applied, the lifecycle:
  - SHALL proceed with the `build` phase using the provided `<gid>` and `<uid>`
- When extending the run image, after all `run.Dockefile`s are applied, the lifecycle:
  - **If** any `run.Dockerfile` set the label `io.buildpacks.rebasable` to `false` or left the label unset:
    - SHALL set the label `io.buildpacks.rebasable` to `false` on the extended run image

>   - SHALL copy the manifest and config for the extended run image, along with any new layers, to `<extended>`/run

Author: ```suggestion
  - **If** after the final `run.Dockerfile` the run image user is `root`,
    - SHALL fail
  - SHALL copy the manifest and config for the extended run image, along with any new layers, to `<extended>`/run
```

- The lifecycle SHALL write the same app image to each `<image>` tag
- The app image:
    - **If** image extensions were used to extend the `run-image` in [`analyzed.toml`](#analyzedtoml-toml):

>       - MUST be based on the image in `<extended>`

Author: ```suggestion
      - MUST be based on the image in `<extended>`/run
```

    - MUST have the working directory set to the value of `<app>`.
    - MUST contain the following labels
        - `io.buildpacks.lifecycle.metadata`: see [lifecycle metadata label](#iobuildpackslifecyclemetadata-json)
        - `io.buildpacks.project.metadata`: the value of which SHALL be the json representation of `<project-metadata>`
        - `io.buildpacks.build.metadata`: see [build metadata](#iobuildpacksbuildmetadata-json)

>         - **If** image extensions were used to extend the run image and `<extended>/<image config>` has the label `io.buildpacks.rebasable` set to `true`:

Author: This might be redundant, as we already say that image config labels are preserved.

    - MUST have the working directory set to the value of `<app>`.
    - MUST contain the following labels
        - `io.buildpacks.lifecycle.metadata`: see [lifecycle metadata label](#iobuildpackslifecyclemetadata-json)
        - `io.buildpacks.project.metadata`: the value of which SHALL be the json representation of `<project-metadata>`
        - `io.buildpacks.build.metadata`: see [build metadata](#iobuildpacksbuildmetadata-json)

>         - **If** image extensions were used to extend the run image and `<extended>/<image config>` has the label `io.buildpacks.rebasable` set to `true`:

Author: ```suggestion
        - **If** image extensions were used to extend the run image and `<extended>/run/<image config>` has the label `io.buildpacks.rebasable` set to `true`:
```

|---------------|----------------------|------------------------|-------------------------------------------------------|
| `<daemon>`    | `CNB_USE_DAEMON`     | `false`                | Export image to docker daemon                         |
| `<force>`     | `CNB_FORCE_REBASE`   | `false`                | Allow unsafe rebase                                   |
| `<gid>`       | `CNB_GROUP_ID`       |                        | Primary GID of the build image `User`                 |
| `<image>`     |                      |                        | App image to rebase                                   |
| `<log-level>` | `CNB_LOG_LEVEL`      | `info`                 | Log Level                                             |
| `<report>`    | `CNB_REPORT_PATH`    | `<layers>/report.toml` | Path to report (see [`report.toml`](#reporttoml-toml) |
| `<run-image>` | `CNB_RUN_IMAGE`      | derived from `<image>` | Run image reference                                   |
| `<uid>`       | `CNB_USER_ID`        |                        | UID of the build image `User`                         |

- At least one `<image>` must be provided

> - **If** `<image>` has `io.buildpacks.rebasable` set to `false`, the lifecycle SHALL fail unless `<force>` is `true`

Author: ```suggestion
- **If** `<image>` has. the label `io.buildpacks.rebasable` set to `false`, the lifecycle SHALL fail unless `<force>` is `true`
```

# layer metadata

[run-image]

>   image = "<image>"

Author: This might be removed in favor of expanding reference to include the tag part (cc @ jjbustamante)

  },
  "stack": {
    "runImage": {
      "image": "cnbs/sample-stack-run:bionic",

>       "mirrors": = ["<mirror1>", "<mirror2>"]

Author: This is part of the label already and was just omitted from the spec
